signer/tests/regtest/mod.rs

```diff
@@ -0,0 +1,363 @@
+use bitcoin::Address;
+use bitcoin::AddressType;
+use bitcoin::Amount;
+use bitcoin::CompressedPublicKey;
+use bitcoin::EcdsaSighashType;
+use bitcoin::Network;
+use bitcoin::OutPoint;
+use bitcoin::PublicKey;
+use bitcoin::ScriptBuf;
+use bitcoin::Sequence;
+use bitcoin::TapLeafHash;
+use bitcoin::TapNodeHash;
+use bitcoin::TapSighashType;
+use bitcoin::Transaction;
+use bitcoin::TxIn;
+use bitcoin::TxOut;
+use bitcoin::Witness;
+use bitcoincore_rpc::json::ImportDescriptors;
+use bitcoincore_rpc::json::ListUnspentQueryOptions;
+use bitcoincore_rpc::json::ListUnspentResultEntry;
+use bitcoincore_rpc::json::Timestamp;
+use bitcoincore_rpc::jsonrpc::error::Error as JsonRpcError;
+use bitcoincore_rpc::jsonrpc::error::RpcError;
+use bitcoincore_rpc::Auth;
+use bitcoincore_rpc::Client;
+use bitcoincore_rpc::Error as BtcRpcError;
+use bitcoincore_rpc::RpcApi;
+use secp256k1::Keypair;
+use secp256k1::Message;
+use secp256k1::SECP256K1;
+use std::sync::OnceLock;
+
+/// These must match the username and password in bitcoin.conf
+const BITCOIN_CORE_RPC_USERNAME: &str = "alice";
+const BITCOIN_CORE_RPC_PASSWORD: &str = "pw";
+
+pub const BITCOIN_CORE_FALLBACK_FEE: Amount = Amount::from_sat(1000);
+
+/// The name of our wallet on bitcoin-core
+const BITCOIN_CORE_WALLET_NAME: &str = "integration-tests-wallet";
+
+/// The faucet has a fixed secret key so that any mined amounts are
+/// preseved between test runs.
+const FAUCET_SECRET_KEY: &str = "0000000000000000000000000000000000000000000000000000000000000001";
+
+const FAUCET_LABEL: Option<&str> = Some("faucet");
+
+static BTC_CLIENT: OnceLock<(Client, Recipient)> = OnceLock::new();
+
+/// Initializes a blockchain and wallet on bitcoin-core. It can be called
+/// multiple times (even concurrently) but only generates the client and
+/// recipient once.
+///
+/// This function does the following:
+/// * Creates an RPC client to bitcoin-core
+/// * Creates or loads a watch-only wallet on bitcoin-core
+/// * Loads a "faucet" private-public key pair with a P2WPKH address.
+/// * Has the bitcoin-core wallet watch the generated address.
+/// * Ensures that the faucet has at least 1 bitcoin spent to its address.
+pub fn initialize_blockchain() -> &'static (Client, Recipient) {
+    BTC_CLIENT.get_or_init(|| {
+        let username = BITCOIN_CORE_RPC_USERNAME.to_string();
+        let password = BITCOIN_CORE_RPC_PASSWORD.to_string();
+        let auth = Auth::UserPass(username, password);
+        let rpc = Client::new("http://localhost:18443", auth).unwrap();
+
+        get_or_create_wallet(&rpc, BITCOIN_CORE_WALLET_NAME);
+        let faucet = Recipient::from_key(FAUCET_SECRET_KEY, AddressType::P2wpkh);
+        faucet.track_address(&rpc, FAUCET_LABEL);
+
+        let amount = rpc
+            .get_received_by_address(&faucet.address, Some(1))
+            .unwrap();
+
+        if amount < Amount::from_int_btc(1) {
+            faucet.generate_blocks(&rpc, 101);
+        }
+
+        (rpc, faucet)
+    })
+}
```

Comment:
I like the pattern in the first example at https://doc.rust-lang.org/std/sync/struct.OnceLock.html where the static variable is local for the getter function, since it should only be read through that function.

```suggestion
/// Initializes a blockchain and wallet on bitcoin-core. It can be called
/// multiple times (even concurrently) but only generates the client and
/// recipient once.
///
/// This function does the following:
/// * Creates an RPC client to bitcoin-core
/// * Creates or loads a watch-only wallet on bitcoin-core
/// * Loads a "faucet" private-public key pair with a P2WPKH address.
/// * Has the bitcoin-core wallet watch the generated address.
/// * Ensures that the faucet has at least 1 bitcoin spent to its address.
pub fn initialize_blockchain() -> &'static (Client, Recipient) {
    static BTC_CLIENT: OnceLock<(Client, Recipient)> = OnceLock::new();
    BTC_CLIENT.get_or_init(|| {
        let username = BITCOIN_CORE_RPC_USERNAME.to_string();
        let password = BITCOIN_CORE_RPC_PASSWORD.to_string();
        let auth = Auth::UserPass(username, password);
        let rpc = Client::new("http://localhost:18443", auth).unwrap();

        get_or_create_wallet(&rpc, BITCOIN_CORE_WALLET_NAME);
        let faucet = Recipient::from_key(FAUCET_SECRET_KEY, AddressType::P2wpkh);
        faucet.track_address(&rpc, FAUCET_LABEL);

        let amount = rpc
            .get_received_by_address(&faucet.address, Some(1))
            .unwrap();

        if amount < Amount::from_int_btc(1) {
            faucet.generate_blocks(&rpc, 101);
        }

        (rpc, faucet)
    })
}
```